src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -999,35 +904,66 @@ CreateHashDistributedTable(Oid relationId, char *distributionColumnName,
 	 */
 	pgDistColocation = heap_open(DistColocationRelationId(), ExclusiveLock);
 
-	/* check for existing colocations */
-	colocationId = ColocationId(shardCount, replicationFactor, distributionColumnType);
+	/* get distribution column data type*/
+	distributionColumnType = ColumnType(relationId, distributionColumnName);
 
-	/*
-	 * If there is a colocation group for the current configuration, get a
-	 * colocated table from the group and use its shards as a reference to
-	 * create new shards. Otherwise, create a new colocation group and create
-	 * shards with the default round robin algorithm.
-	 */
-	if (colocationId != INVALID_COLOCATION_ID)
+	if (pg_strncasecmp(colocateWithTableName, "default", NAMEDATALEN) == 0)
 	{
-		char *relationName = get_rel_name(relationId);
+		/* check for default colocation group */
+		colocationId = DefaultColocationGroupId(shardCount, replicationFactor,
+												distributionColumnType);
+		if (colocationId == INVALID_COLOCATION_ID)
+		{
+			bool defaultColocationGroup = true;
+			colocationId = CreateColocationGroup(shardCount, replicationFactor,
+												 distributionColumnType,
+												 defaultColocationGroup);
+		}
+		else
+		{
+			colocationTableId = ColocatedTableId(colocationId);
+		}
+	}
+	else if (pg_strncasecmp(colocateWithTableName, "none", NAMEDATALEN) == 0)
+	{
+		bool defaultColocationGroup = false;
+		colocationId = CreateColocationGroup(shardCount, replicationFactor,
+											 distributionColumnType,
+											 defaultColocationGroup);
+	}
+	else
+	{
+		Var *colocationTablePartitionColumn = NULL;
+		Oid colocationTablePartitionColumnType = InvalidOid;
 
-		Oid colocatedTableId = ColocatedTableId(colocationId);
-		ConvertToDistributedTable(relationId, distributionColumnName,
-								  DISTRIBUTE_BY_HASH, colocationId);
+		/* get colocation group of the target table */
+		text *colocateWithTableNameText = cstring_to_text(colocateWithTableName);
+		colocationTableId = ResolveRelationId(colocateWithTableNameText);
+
+		colocationId = TableColocationId(colocationTableId);
+
+		colocationTablePartitionColumn = PartitionKey(colocationTableId);
+		colocationTablePartitionColumnType = colocationTablePartitionColumn->vartype;
 
-		CreateColocatedShards(relationId, colocatedTableId);
-		ereport(DEBUG2, (errmsg("table %s is added to colocation group: %d",
-								relationName, colocationId)));
+		if (colocationTablePartitionColumnType != distributionColumnType)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot colocate with %s", colocateWithTableName),
+							errdetail("Distribution column types are different.")));
+		}
```

Comment:
I'm slightly inclined not to do this because we are trying to hide internals from the user including shard count. In an ideal setup, the user shouldn't care about the shard count, and the power user should already know about the defaults etc.